### ISSUE TITLE
[Ruby] Improve percent literals highlighting

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#124]
https://docs.ruby-lang.org/en/master/syntax/literals_rdoc.html#label-Percent+Literals
https://ruby-doc.org/3.2.0/syntax/literals_rdoc.html#label-Percent+Literals

**Comment by nyamatongwe** [comment]
The `ruby-*.org` links don't actually show any `%s` examples.

Unlike other quoted literal types `SCE_RB_STRING_Q`..., there is no lexical state for symbol literals. `s` is also not included in potential quoting characters `q_chars`.

**Comment by zufuliu** [comment]
Yeah, it's missing from the source at https://github.com/ruby/ruby/blob/master/doc/syntax/literals.rdoc#label-25s-3A+Symbol+Literals. I think it can be handled similar to other percents.

**Comment by zufuliu** [comment]
I'd like to change LexRuby.cxx to use anonymous namespace before adding extra four styles (`%w`, `%i`, `%I`, `%s`) to implement other percent literals.

**Comment by zufuliu** [comment]
[LexRuby-0210.zip](https://github.com/ScintillaOrg/lexilla/files/10702261/LexRuby-0210.zip)

change LexRuby.cxx to use anonymous namespace and `nullptr`, add `const` and `noexcept`, removed assignment inside `if`.

**Comment by zufuliu** [comment]
This (and the two linked issues) can be fixed by adding new styles for percent literals, as following table:

| literal | style | interpolable | remarks |
|-|-|-|-|
| `%q` | SCE_RB_STRING_Q | ❌| string |
| `%Q` | SCE_RB_STRING_QQ | ✅| string |
| `%r` | SCE_RB_STRING_QR | ✅| regex |
| `%w` | SCE_RB_STRING_W | ❌| string array, new style, currently handled same as `%W` |
| `%W` | SCE_RB_STRING_QW | ✅| string array |
| `%x` | SCE_RB_STRING_QX | ✅| backticks |
| `%i` | SCE_RB_STRING_I | ❌| symbol array, new style |
| `%I` | SCE_RB_STRING_QI | ✅| symbol array, new style |
| `%s` | SCE_RB_STRING_QS | ❌| symbol, new style |

the style name conversion is: for letter `<?>`, if there is only single case then name the style as `SCE_RB_STRING_Q<?>`, e.g. SCE_RB_STRING_QR, SCE_RB_STRING_QX; else name the interpolable one as `SCE_RB_STRING_Q<?>` and non-interpolable one as `SCE_RB_STRING_<?>`.

**Comment by nyamatongwe** [comment]
Removing "static" from function definitions moves the "(" and first argument which means that further argument lines don't align as they did before. Its easier to do this aligning with astyle. The common formatting of `LexRuby.cxx` matches Scintilla except for using spaces instead of tabs so this command was used for 955c79d87e837c33edb7af21c385acf7d993d040:
~~~
astyle --style=attach --indent=spaces=4 --keep-one-line-blocks
   --pad-header --unpad-paren --pad-comma --indent-cases --align-pointer=name --pad-method-prefix
   --pad-return-type --pad-param-type --align-method-colon --pad-method-colon=after
~~~

Initial reformatting changed the position of 7 braces.

The only element not copied from `LexRuby-0210.zip` was the change of type for `q_states` from `int` to `unsigned char` as its only use was to assign to an `int` so the change mixed types without need.

Reordered includes to match `scripts/HeaderOrder.txt`.

**Comment by zufuliu** [comment]
[ruby-0211.zip](https://github.com/ScintillaOrg/lexilla/files/10712071/ruby-0211.zip)

Added four styles, test case added into AllStyles.rb.

**Comment by zufuliu** [comment]
Following loop can be removed as the two arrays already zeroed, the code does nothing in release build.
https://github.com/ScintillaOrg/lexilla/blob/b62dcea558597e1ba634f73451977b4ca9332c22/lexers/LexRuby.cxx#L784-L796

**Comment by zufuliu** [comment]
[ruby-0211-2.zip](https://github.com/ScintillaOrg/lexilla/files/10712765/ruby-0211-2.zip)

Forget to add code folding for SCE_RB_STRING_W.
```diff
-        } else if (style == SCE_RB_STRING_QW) {
+        } else if (style == SCE_RB_STRING_QW || style == SCE_RB_STRING_W) {
```

**Comment by zufuliu** [comment]
https://sourceforge.net/p/scintilla/bugs/1255/ (`%=`) can be fixed with following change:
```diff
@@ -1082,7 +1082,7 @@ void ColouriseRbDoc(Sci_PositionU startPos, Sci_Position length, int initStyle,
                     Quote.Open(chNext);
                     advance_char(i, ch, chNext, chNext2); // pass by ref
                     have_string = true;
-                } else if (!isSafeWordcharOrHigh(chNext) && !iswhitespace(chNext) && !isEOLChar(chNext)) {
+                } else if (!isSafeWordcharOrHigh(chNext) && !iswhitespace(chNext) && !isEOLChar(chNext) && chNext != '=') {
                     // Ruby doesn't allow high bit chars here,
                     // but the editor host might
                     Quote.New();
```

for following Ruby code:
```ruby
def hello(what)
	return %=#{what}=
end

puts hello(%=1=)
puts %=2=
```

first two `%=` care covered by `preferRE`, but last line is invalid:
```
ruby t1.rb
t1.rb: --> t1.rb
syntax error, unexpected '=', expecting end-of-input
> 5  puts hello(%=1=)
> 6  puts %=2=
t1.rb:6: syntax error, unexpected '=', expecting end-of-input (SyntaxError)
puts %=2=
        ^
```

[ruby-0214.zip](https://github.com/ScintillaOrg/lexilla/files/10731734/ruby-0214.zip)
Patch merged the two `// Ruby doesn't allow high bit chars here` blocks, replaced `iswhitespace` with `IsASpaceOrTab`, added `isWhiteSpace` to test space tab and eols.

**Comment by nyamatongwe** [comment]
Included in 5.2.3.
